### PR TITLE
fix MAXIMUM_NUMBER_OF_STREAMS errors when doing many websocket queries

### LINKED ISSUE #15551: MAXIMUM_NUMBER_OF_STREAMS thrown even though nr of active clients is low
# Affected Daml version

2.5.0-snapshot.20221107.10906.0.346bb489 (possibly earlier ones as well, but I've not seen it with 2.5.0-snapshot.20221010.10736.0.2f453a14)

# Bug description

After running and using the `daml-finance-app` for a while I'm getting errors relating to websocket streams:

```
11-11-2022 14:55:32.225 [http-json-ledger-api-akka.actor.default-dispatcher-20] INFO  com.daml.http.WebSocketService - Websocket client interrupted on Failure: ABORTED: MAXIMUM_NUMBER_OF_STREAMS(2,0): The number of streams in use (1000) has reached or exceeded the limit (1000). Metrics are available at daml.lapi.streams.active.. remaining number of clients: 39, context: {instance_uuid: "36983077-5bda-45cf-85b7-81081f8a825e", request_id: "cfdd9502-2cc9-4c85-86fd-09b95a83fbac"} 
```

I'm also occasionally seeing this one, but not sure it's related:

```11-11-2022 14:58:35.619 [http-json-ledger-api-akka.actor.default-dispatcher-7] INFO  com.daml.http.WebSocketService - Websocket client interrupted on Failure: ABORTED: THREADPOOL_OVERLOADED(2,0): The Index Database Connection Threadpool queue size (1005) has exceeded the maximum (1000). Api services metrics are available at daml.index.db.threadpool.connection.api-server.. remaining number of clients: 16, context: {instance_uuid: "3a81ec97-874f-4d06-953f-2651aca513d4", request_id: "7c087d49-a4d6-476c-9c14-8e4ae7473a4b"} 
11-11-2022 14:58:35.623 [http-json-ledger-api-akka.actor.default-dispatcher-13] ERROR akka.actor.ActorSystemImpl - Websocket handler failed with ABORTED: THREADPOOL_OVERLOADED(2,0): The Index Database Connection Threadpool queue size (1005) has exceeded the maximum (1000). Api services metrics are available at daml.index.db.threadpool.connection.api-server. 
io.grpc.StatusRuntimeException: ABORTED: THREADPOOL_OVERLOADED(2,0): The Index Database Connection Threadpool queue size (1005) has exceeded the maximum (1000). Api services metrics are available at daml.index.db.threadpool.connection.api-server.
        at io.grpc.Status.asRuntimeException(Status.java:535)
        at io.grpc.stub.ClientCalls$StreamObserverToCallListenerAdapter.onClose(ClientCalls.java:479)
        at io.grpc.internal.ClientCallImpl.closeObserver(ClientCallImpl.java:562)
        at io.grpc.internal.ClientCallImpl.access$300(ClientCallImpl.java:70)
        at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInternal(ClientCallImpl.java:743)
        at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInContext(ClientCallImpl.java:722)
        at io.grpc.internal.ContextRunnable.run(ContextRunnable.java:37)
        at io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:133)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
        at java.base/java.lang.Thread.run(Thread.java:829)
```

The app gets unresponsive after seeing the first error, so it seems that it's not just wrong reporting of nr of streams.

I'm not sure if this was already present in a previous snapshot, I've only seen it happen now on the latest (version above).

# To reproduce

- Clone [daml-finance-app](https://github.com/digital-asset/daml-finance-app)
- Run it as described [here](https://github.com/digital-asset/daml-finance-app#running)
- load `localhost:3000` in 3 or 4 tabs, quickly, until the error appears in the `daml start` log
,
# Expected behavior

As said, this seems to be related to some recent changes, as I've never seen this on older snapshots or <= 2.4.x

### PR DESCRIPTION
**Pull request by S11001001**
Fixes #15551. This fixes `MAXIMUM_NUMBER_OF_STREAMS` errors with Canton 2.5.0+ when using websocket query; ACS and transaction stream management would not previously clean up these streams properly on closure of the websocket.

Several log messages are included for future diagnosis if necessary; to see them, set the log-level of `com.daml.fetchcontracts.util.GraphExtensions` to `TRACE` and grep for `stream-`.

As a minor side matter, this also fixes the `current number of clients` log messages to be slightly more accurate.

```sh
bazel test -k //ledger-service/fetch-contracts:tests
```

### CODE REVIEW

**Comment by S11001001** [comment]
Regarding 52a027f840b8e7d0ce7295058f2cad470830e903: 2.6.20 does not change the implementation of `flatMapMerge` (we are on 2.6.18), so don't bother with that approach.

File: ledger-service/http-json/src/main/scala/com/digitalasset/http/WebSocketService.scala [hunk 1]
```diff
@@ -853,13 +853,20 @@ class WebSocketService(
             jwtPayload.parties,
             offPrefix,
             rq.q: q,
-          )
+          ).watchTermination() { (m, fd) =>
+            fd.onComplete(_ => logger.info("S11 finished in FMM"))
+            m
+          }
```
Comment (ray-roestenburg-da) [inline ledger-service/http-json/src/main/scala/com/digitalasset/http/WebSocketService.scala:859]: In the docs:

> Completes when upstream completes and all consumed substreams complete
> Cancels when downstream cancels

When would `getTransactionSourceForParty` complete? maybe what is necessary is to cancel the sources on completion of the `wsMessageHandler`.

**Comment by S11001001** [comment]
Confirmed that restarting json-api itself without restarting sandbox immediately recovers the system in local testing.

File: ledger-service/fetch-contracts/src/main/scala/fetchcontracts/AcsTxStreams.scala [hunk 2]
```diff
@@ -110,12 +117,14 @@ private[daml] object AcsTxStreams {
...
       val maxOff = b add max(LedgerBegin: Off)
+      val logTxnOut = b add Flow[ContractStreamStep.Txn.LAV1].logTermination("after-split")
       // format: off
       discard { txnSplit.in <~ txns <~ dupOff }
       discard {                        dupOff                                ~> mergeOff ~> maxOff }
       discard { txnSplit.out1.map(off => AbsoluteBookmark(off)) ~> lastTxOff ~> mergeOff }
+      discard { txnSplit.out0 ~> logTxnOut }
       // format: on
-      new FanOutShape2(dupOff.in, txnSplit.out0, maxOff.out)
+      new FanOutShape2(dupOff.in, logTxnOut.out, maxOff.out)
```
Comment (S11001001) [inline ledger-service/fetch-contracts/src/main/scala/fetchcontracts/AcsTxStreams.scala:127]: @ray-roestenburg-da Should the final PR leave all this `logTermination` spam everywhere, maybe set to `trace` or `debug` (and with `S11` removed of course)?

Comment (ray-roestenburg-da) [reply ledger-service/fetch-contracts/src/main/scala/fetchcontracts/AcsTxStreams.scala:127]: yes, sounds good.

**Review by ray-roestenburg-da — APPROVED** [review]

**Pull request merged**